lib/src/entrypoint.dart

```diff
@@ -916,6 +931,113 @@ See https://dart.dev/go/sdk-constraint
     dataError('The "$packageConfigPath" file is not recognized by '
         '"pub" version, please run "$topLevelProgram pub get".');
   }
+
+  /// Looks through the root-[incompability] of a solve-failure and tries to see
+  /// if any single change can make a resolution go through. Returns a formatted
+  /// list of suggestions, or the empty String if no suggestions were found.
+  Future<String> _suggestResolutionAlternatives(SolveType type,
+      Incompatibility incompatibility, Iterable<String> unlock) async {
+    final b = StringBuffer();
+    final visited = <String>{};
+    final stopwatch = Stopwatch()..start();
+    var first = true;
+    void suggest(String suggestion) {
+      if (first) {
+        b.writeln(
+            '\nYou can try one of the following suggestion(s) to resolve the failure:');
+      }
+      b.writeln(suggestion);
+      first = false;
+    }
+
+    for (final externalIncompatibility
+        in incompatibility.externalIncompatibilities) {
+      if (stopwatch.elapsed > Duration(seconds: 3)) {
+        // Never spend more than 3 seconds computing suggestions.
+        break;
+      }
+      final cause = externalIncompatibility.cause;
+      if (cause is SdkCause) {
+        final sdkName = cause.sdk.identifier;
+        if (!(sdkName == 'dart' ||
+            (sdkName == 'flutter' && runningFromFlutter))) {
+          // Only make sdk upgrade suggestions for Flutter and Dart.
+          continue;
+        }
+
+        final constraint = cause.constraint;
+        if (constraint == null) continue;
+
+        /// Find the most relevant Flutter release fullfilling the constraint.
+        final bestRelease =
+            await inferBestFlutterRelease({cause.sdk.identifier: constraint});
+        if (bestRelease == null) continue;
+        try {
+          await resolveVersions(type, cache, root,
+              lockFile: lockFile,
+              unlock: unlock,
+              sdkOverrides: {
+                'dart': bestRelease.dartVersion,
+                'flutter': bestRelease.flutterVersion
+              });
+
+          if (runningFromFlutter) {
+            suggest(
+                '* Try using the Flutter SDK version: ${bestRelease.flutterVersion}. ');
+          } else {
+            // Here we assume that any Dart version included in a Flutter
+            // release can also be found as a released Dart SDK.
+            suggest(
+                '* Try using the Dart SDK version: ${bestRelease.dartVersion}. See https://dart.dev/get-dart.');
```

Comment:
Can we say "consider" whenever we're suggesting a downgrade?

Reply:
Done